error.go

```diff
@@ -124,6 +124,33 @@ func (e *Err) Error() string {
 	return fmt.Sprintf("%s: %v", e.message, err)
 }
 
+// Format implements fmt.Formatter
+// When printing errors with %+v it also prints the stack trace.
+// %#v unsurprisingly will print the real underlying type.
+func (e *Err) Format(s fmt.State, verb rune) {
+	switch verb {
+	case 'v':
+		switch {
+		case s.Flag('+'):
+			fmt.Fprintf(s, "%s", ErrorStack(e))
+			return
+		case s.Flag('#'):
+			// avoid infinite recursion by wrapping e into a type
+			// that doesn't implement Formatter.
+			fmt.Fprintf(s, "%#v", (*unformatter)(e))
```

Comment:
no strong opinions here; my expectations for `%#v` was to show the actual thing without any smart reformatting, i.e. as the doc says:

```
%#v	a Go-syntax representation of the value
```

So currently, with gopkg.in/check.v1 you see the guts of juju/errors:



```
	err := fmt.Errorf("foo")
	c.Check(err, Equals, 42)
	err = errors.Trace(err)
	c.Check(err, Equals, 42)

... go test:

foo_test.go:21:
    c.Check(err, Equals, 42)
... obtained *errors.errorString = &errors.errorString{s:"foo"} ("foo")
... expected int = 42

foo_test.go:23:
    c.Check(err, Equals, 42)
... obtained *errors.Err = &errors.Err{message:"", cause:(*errors.errorString)(0xc42000ef60), previous:(*errors.errorString)(0xc42000ef60), file:"pippo/foo_test.go", line:22} ("foo")
... expected int = 42
```

With this PR, the latter will become:

```
&errors.unformatter{message:"", cause:(*errors.errorString)(0xc42000ef60), previous:(*errors.errorString)(0xc42000ef60), file:"pippo/foo_test.go", line:22}
```

IMHO it's a bit ugly but at least it's not hiding anything.

@rogpeppe: Still prefer `errors.Details(e)` ?